### ISSUE TITLE
Does a client instance really need to be identified by a unique key ?

### ISSUE THREAD

**Issue by Denisthemalice** [ietf-wg-gnap/gnap-core-protocol#292]
On page 6, the text states:

> Note: this specification differentiates between a specific instance (the client instance, **_identified by its unique key_**) and the software running the instance (the client software). For some kinds of client software, there could be many instances of that software, each instance with a different key.

Section 2.3 is called: 

> Identifying the Client Instance

It is quite odd to say that a **_unique key_** identifies a client instance. 
The wording “unique key“ does not appear anywhere else in the document.

On page 23, in section 2 (Requesting Access), the text describes the following data element:

> client (object / string) Describes the client instance that is making this request, including the key that the client instance will use to protect this request and any continuation requests at the AS and any user-facing information about the client instance used in interactions.

Protecting the exchanges between the client and the AS using a unique key specific to a client looks like over-engineering. 
In most cases, a given user uses a client instance, so mandating to protect the exchanges using HTTPS is straightforward and sufficient. 

Later on in the document (page 80), it appears that two options are possible to bind an access token to a key:

> If the "flags" field does not contain the "bearer" flag and the "key" is absent, the access token MUST be sent using the same key and proofing mechanism that the client instance used in its initial request (or its most recent rotation).
> 
> If the "flags" field does not contain the "bearer" flag and the "key" value is an object as described in Section 7.1, the access token MUST be sent using the key and proofing mechanism defined by the value of the "proof" field within the key object.

Supporting only the second option and mandating the use of HTTPS between a client instance and an AS would lead to simplifications.

**Comment by jricher** [comment]
The key is the core identity of the client instance. Merely requiring https would not allow differentiation between instances nor authentication nor registration.

**Comment by Denisthemalice** [comment]
@jricher.  Would you be able to explain what you mean by:
> Merely requiring https would not allow differentiation between instances nor authentication nor registration. ?

**Comment by jricher** [comment]
If two different client instances make the same request over HTTPS, how does the AS differentiate between them? How are subsequent requests protected? HTTPS does nothing here.

**Comment by Denisthemalice** [comment]
I believed I had responded to this argument, but apparently it is not the case (or I have responded in another issue).

Section 7 (Securing Requests from the Client Instance) states:

> “In GNAP, the client instance secures its requests to the AS (…) by presenting an access token, presenting proof of a key 
that it possesses, or both an access  token and key proof together”.

The Security Considerations section states:

> All requests have to be over TLS or equivalent as per [BCP195].

If the authentication exchange between the user and the AS is protected using TLS, then it is no more necessary to manage client instance keys. This would allow major simplifications.

**Comment by aaronpk** [comment]
Client instances are identified by keys so that access tokens issued to them can be restricted to be only usable by the specific client instance with that key. One of the core limitations with cookies and bearer tokens is that anyone that can manage to get one can use it.

An analogous real-world example is a hotel key. If you find a hotel key on the floor, you can pick it up and go try it at the door and the door has no idea whether it’s the actual guest or someone else who is using it. (In this example, the person using the hotel key is analogous to the client software.) A hotel key is like a bearer token. 

Ideally we would like to do better in terms of security, such that the access tokens issued to clients are usable only by the particular client instance they were issued to.

The most secure way of doing that is by using public/private keys to identify and authenticate the client, ideally in a hardware security module, so that the private key can’t be extracted from the client.

**Comment by Denisthemalice** [comment]
The current specification does not mandate the use of hardware security modules for the clients. If such an hardware security module were used, which would have the single security property to prevent the private key to be extracted out of it, this would not provide any kind of security since the legitimate client could cooperate with another client and perform all the cryptographic computations that the other client would need.

This means that access tokens issued to client instances cannot be restricted to be only usable by the specific client instance 
with that key, even if hardware security modules were used.

In order to secure the exchanges, there is a need to make a difference between the exchanges between the client and the **_AS_** 
and those between the client and the **_RS_**.

**Security of the client-AS exchanges** 

If the authentication exchange between the user and the AS is protected using TLS, the access token request made through the same TLS channel originates from the user that has just been successfully authenticated. This does not require the use of client instance keys and hence this would allow major simplifications.

**Security of the client-RS exchanges**

For this case, there is no authentication exchange between the user and the RS. A solution is to include a field in the access token 
that allows an RS to verify that the access token is intended for the right user. For more details, refer to the issue #290

As a conclusion, in both cases, **_client instance keys are not needed_**.

**Comment by aaronpk** [comment]
> If the authentication exchange between the user and the AS is protected using TLS, the access token request made through the same TLS channel originates from the user that has just been successfully authenticated

This sentence assumes the use of the same TLS channel, which is completely not how any of this works. It is quite common to use new TLS connections for subsequent requests to the AS, for example refreshing an access token hours or days later.

> A solution is to include a field in the access token that allows an RS to verify that the access token is intended for the right user

You have again failed to state how the RS is supposed to know what user to expect in order to verify the field in the access token matches the expected value. This is already being discussed in #290.

**Comment by Denisthemalice** [comment]
You wrote:

> This sentence assumes the use of the same TLS channel, which is completely not how any of this works. 

It is obvious that what I am proposing is "completely not how any of this works" as described in the current draft.

You also wrote :

> It is quite common to use new TLS connections for subsequent requests to the AS, for example refreshing an access token 
hours or days later.

It is quite common to ask for a user re-authentication if the TSL session with the AS is during too long. 
In many systems, this is done automatically if a user inactivity is detected during, e.g. 10 or 20 minutes at most.

In addition, it is a good practice to have access token duration less than 12 hours and hence there is no need "refreshing an access token hours or days later". When an access token expires, a new access token should be requested (_if the TSL channel with the RS is still open, which is very doubtful_).

I understand that this is not what is being described in the current document, but the new approach I propose would lead to many simplifications. 

Finally, you wrote :

> You have again failed to state how the RS is supposed to know what user to expect in order to verify the field in the access token matches the expected value. This is already being discussed in #290.

This is an argument that relates to another issue. I have not "_failed_" to state how the RS is supposed to know what user to expect in order to verify the field in the access token matches the expected value, I have not yet explained it. There are different cases to consider which depend on how a user account is created. I will come back with a response to your question on this topic in issue #290 (which is where you should have raised your interesting question).

**Comment by aaronpk** [comment]
> It is quite common to ask for a user re-authentication if the TSL session with the AS is during too long.
> In many systems, this is done automatically if a user inactivity is detected during, e.g. 10 or 20 minutes at most.

I am glad we are getting to the bottom of what you are suggesting. I haven't seen any prior comments from you about this idea, but it sounds like you are suggesting that token lifetimes and other policies are tied to individual TLS sessions. If that is what you're suggesting, there are quite a number of implications of a change like that.

**Comment by aaronpk** [comment]
@Denisthemalice client instance keys are needed so that the AS can associate a particular access token with a client instance key, such that the access token is only usable by the individual client instance because they are forced to prove possession of the key when using the access token.

the attacks this prevents are all in the family of stolen access tokens. For example a malicious browser extension is able to extract a token from the browser's LocalStorage and send it off to the attacker's remote server. With client instance keys, the attacker is unable to use the token because they would not have been able to also extract the private key from the browser.

**Comment by Denisthemalice** [comment]
You indicated that the goal is to have the access token only usable by the individual **_client instance_**. 
The right goal is to have the access token only usable by a particular **_user._**

It is important to make a difference between **_clients_** and **_users_**.

The use of client instances keys does not prevent two client instances to collaborate, without the need "to extract the private key from the browser". The collaborative client can perform all the computations that the other collaborative client needs. Protecting private keys in a security module or in a secure element would not help either. 

Binding an access token to a particular client instance key, does not prevent it to be forwarded to another user.

**_Client instances keys_** are also used for the exchanges between the client and the AS. 

In order to bind the access token request (and other requests) with a **_user_**, it should first be observed that the use of TLS allows 
the AS to make sure that all the messages are coming from a single **_unknown_** client, 

**_If the authentication exchange between the user and the AS is first conveyed using the TLS channel established between the Client and the AS_**, then all subsequent messages are necessarily coming from that authenticated user.

A simple example is to send, through the TLS channel mounted between the Client and the AS, a UserID and a (long) password followed by an access token request. The access token request necessarily comes from the authenticated user. 

Since the returned access token will necessarily include a "**_bid_**" field  (**_Binding User Identifier_**) it will take benefits of the mechanism described in issue #290.    

In issue #290 you mentioned that client instances keys are preventing an **unrelated attack** to the issue being discussed here.
**What is exactly this unrelated attack ?** 

Are you thinking of an attack that would not be defeated using a combination of  the "**_bid_**" field  (**_Binding User Identifier_**) placed 
in the access token and the support of the authentication exchange inside the TLS channel mounted between the Client and the AS ?

**Comment by jricher** [comment]
@Denisthemalice the use of client keys prevents two non-collaborating client instances from attacking each other. A stolen on-the-wire credential, such as an access token, becomes useless without it. It also prevents an attacker from easily pretending to be a different piece of software to trick a user into authorizing them. TLS offers none of these protections on its own because it protects the channel integrity, not the application integrity between multiple parties.

**Comment by aaronpk** [comment]
> You indicated that the goal is to have the access token only usable by the individual client instance. The right goal is to have the access token only usable by a particular user.

you actually need both. It turns out users don't make API requests. Users use software to make API requests. You need access tokens useable by a particular user tied to a particular client.

**Comment by Denisthemalice** [comment]
@aaronpk :
> It turns out users don't make API requests. Users use software to make API requests. You need access tokens useable by a particular user tied to a particular client.

You need access tokens useable by a particular user where the user is using a particular client.
To obtain this property, it is unnecessary to authenticate the client instance, hence no key needs to be attached to a client instance.

At the moment, the core-document states nothing about **_user authentication_**. 

In section 1.4 (Sequences), Step (2) is described as:

> (2) The client instance requests access at the AS 

The following sections 1.4.1 to 1.4.3 (which are only **_EXAMPLES_**), nothing is said about user authentication.

Section 1.4.4 states:

> the AS policy allows the client instance to  make a call on its own behalf,

This is in contradiction with the two following definitions:
> End-user  : **_natural person_** that operates a client instance.

and also:
> Client :  application operated by an **end-user** 



Section 1.4.5 (still an _**EXAMPLE**_) states nothing about user authentication.

**_There is a link missing in the description : how user authentication is linked to the right client instance._**

We need an access token request linked to a particular TLS channel. In order to get this link, that TSL channel is first established by the client, then the user is invited to authenticate to the AS by choosing among a set of authentication methods proposed by the AS and then after API requests can be done by the client application to the AS using that TLS channel. 

@jricher. You wrote:
> TLS offers none of these protections on its own because it protects the channel integrity, not the application integrity 
between multiple parties.

TLS offers both channel integrity and channel confidentiality. 

- The client can be confident that data coming from the TLS channel are indeed coming from the AS 
(otherwise it knows that it has been corrupted). 
- The AS can be confident that data coming from the TLS channel are indeed coming from unauthenticated client, 
if an authentication of the user has successfully be done by the AS through that TLS channel, and if further API requests 
are indeed coming from the client application through the TLS channel used by the authenticated user.

**Comment by jricher** [comment]
The lines you quote are not in contradiction with each other: the client instance can be acting on its own behalf regardless of who is using it at a given time. The end user need not be the entity that authorizes the client -- they can be, and often are, but that's not the only case as shown in the sections that you point to. There is no conflict just because there are different possibilities.

As for TLS it sounds like you are asking for a protocol where the end user authenticates by presenting their credentials to the client and having the client present those credentials to the AS. Otherwise, how would the user authenticate "through the TLS channel"? Moving away from this password-sharing anti-pattern is one of the foundational goals of delegation protocols: an end user should be able to delegate to software without having to share their credentials with that software.

**Comment by Denisthemalice** [comment]
You wrote:

>The lines you quote are not in contradiction with each other: the client instance can be acting on its own behalf regardless of who is using it at a given time. The end user need not be the entity that authorizes the client -- they can be, and often are, but that's not the only case as shown in the sections that you point to. There is no conflict just because there are different possibilities.

See my detailed response in issue #290.

You wrote:

> As for TLS it sounds like you are asking for a protocol where the end user authenticates by presenting their credentials to the client and having the client present those credentials to the AS. 

This is correct.

Then you wrote:

> Moving away from this password-sharing anti-pattern is one of the foundational goals of delegation protocols: 
an end user should be able to delegate to software without having to share their credentials with that software.

In this architecture, the goal is to use **_access tokens between clients and RSs_**, but a some point of time, there is a need to bootstrap the system. The simplest way to bootstrap the system is to use an **_authentication exchange between the user and the AS_**.

Sharing a password (or using any other authentication method) with an AS is fine, whereas this would not be the case 
with numerous ASs.

Let me recall one of my previous arguments that has not been responded:

> **_There is a link missing in the description : how user authentication is linked to the right client instance._**

**Comment by jricher** [comment]
Let me say that another way: Removing user authentication from the view of the client is a foundational goal of this protocol. That's why it is not mentioned in the places you're looking for it -- it's antithetical to the purpose. 

It sounds to me like you want a different protocol, for a different use case, with a different architecture, with a different goal -- all very separate from what this group has been chartered to work on and the expectations of those engaging in this group's work. To make better progress on your ideas, I would personally recommend that you write your proposal as a complete and comprehensive independent draft, submit it to the IETF for dispatch, and perhaps start a new working group.

**Comment by Denisthemalice** [comment]
I understand that you acknowledge the fact that the current document does not explain how user authentication is linked 
to the right client instance. This missing link is not even advertised in the draft. 

If **_fully removing user authentication_** is indeed **_a foundational goal of this protocol_** then the document should state it
and explain how this can be obtained.

If you believe that the user should present to the AS an access token obtained from another AS, then  you are simply pushing 
the same problem to that other AS and hence you are not solving the problem. **The architecture needs to be "bootstrapable"**.

Having one user authentication exchange with the AS and no user authentication exchange at all with the RSs seems acceptable to me.

**Comment by jricher** [comment]
@Denisthemalice I said none of those things.

**Comment by fimbault** [comment]
Again here, since you asked that I read your 2 issues before our talk, I read this issue in detail. It's not clear what we'd gain from removing key bindings, nor what the alternative would look like.

**Comment by fimbault** [comment]
ok, i've had Denis over the phone, so I think I better understand what he's proposing. 

He's making the valid observation that managing keys is complex. For instance, it's not trivial to distribute and rotate keys between several devices (your pc, tablet, phone, etc.). 
His idea would be to establish authenticated channels (for instance TLS, but we could reuse ideas such as [ockam channels](https://www.ockam.io/learn/concepts/secure_channels) for a more generic implementation).

**Comment by aaronpk** [comment]
This isn't something the user is expected to do, which is why it's called the client instance key, not user key. It would be completely transparent to users.

**Comment by fimbault** [comment]
I would say mostly transparent to users. There are always edge cases when that's not the case (for instance when moving to a new device).

**Comment by aaronpk** [comment]
When moving to a new device, the device would have a new key, so they would end up needing to reauthorize the new device. There wouldn't be any process by which the user would be expected to move keys around, because ideally these devices would be using keys in hardware security modules so the keys couldn't be extracted anyway.

**Comment by jricher** [comment]
Completely agree with @aaronpk that a new device would mean a new instance would mean a new key. The system could allow some kind of smart bootstrapping process so that the authorization process is smoother, but that's outside of GNAP's core concerns.

Relying on TLS channels alone doesn't even get you past a second socket connection, unless you're doing some level of token binding which requires special support at the lower layers and is not widely supported or deployed.

**Comment by fimbault** [comment]
Just saying that the bootstrapping phase in particular may impact significant the user experience here. Sure it's not part of GNAP itself, but still worth considering in the broader picture. 

That being said, it's more to explain what I gathered from Denis, I'm not sold on the idea of secure channels which are really really hard to implement correctly. What is currently proposed in gnap seems way more practical to me. Just so that we're clear.

**Comment by Denisthemalice** [comment]
What is not explained in the current specification,  is that each user must contact the AS in order to be first associated with one client instance key. (Thanks Fabien for that explanation !)

If a user has four devices (e.g. a desktop, a laptop, a tablet and a smartphone) and wants to use each of them, he needs to ask to the AS four client instance keys. If it looses his tablet (or his tablet is stolen), he needs to ask to the AS to revoke the key associated with the tablet and then after when he gets a new tablet, he needs to be associated with a new tablet client instance key. 

As Fabien indicated: 
>  managing keys is complex. For instance, it's not trivial to distribute and rotate keys between several devices (your pc, tablet, phone, etc.). 

This key management can be avoided by using a different approach. That approach allows the user to use any of his four devices (e.g. a desktop, a laptop, a tablet and a smartphone) without the need to register any client instance key with the AS. If the user is going on vacations, he can pick up one of his mobile devices and he knows that he will be able to access to his usual resources. He can even forget to take any of his devices and buy a new one when he arrives at his vacancy location or even use the device of someone else if it believes that this device is "_well managed_". Such functionality is called **_Ease of Use_**.

The main idea is to use the security properties of TLS sessions between (1) the Client and the AS and (2) the Client and the RS.
Once a Client has successfully opened a TLS session with an AS, the exchanges are both integrity and confidentiality protected, the client knows that data coming through that TLS session are indeed coming from the AS, while the AS knows that data coming trough that TLS session are indeed coming from an **_unknown client_**.

That channel is asymmetric and the idea is NOT to reuse ideas such as ockam channels (as Fabien proposed) which provide bi-lateral authentication and hence are less easy to manage when compared to TLS. 

The following of the scenario is then rather simple. Once the TLS channel has been opened, the User authenticates towards the AS through that TLS channel (he can even use an Id/password authentication scheme). At this point of time, if the authentication exchange has been successful, the AS can be confident that an authenticated User is indeed present. Subsequent APIs can then be sent through that TLS channel, which means that the AS can be confident that they originate from the authenticated User. The prime example of such API is an access token request which will indicate (among other parameters) which type of **_"bui"_** field **_(binding user identifier)_** should be included into the requested access token.

This approach allows to secure the exchanges between (1) the Client and the AS and (2) the Client and the RS without the burden of a client instance key management system.

**Comment by aaronpk** [comment]
> What is not explained in the current specification, is that each user must contact the AS in order to be first associated with one client instance key.

From the [introduction](https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-06.html#name-introduction): "This protocol allows a piece of software, the client instance, to request delegated authorization to resource servers", and the term "client instance" is described in the section [defining the roles](https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-06.html#section-1.2-3.4.1).

This seems to describe the situation accurately to me. Do you have a suggestion for making this clearer?

> That approach allows the user to use any of his four devices (e.g. a desktop, a laptop, a tablet and a smartphone) without the need to register any client instance key with the AS

How do you propose this sharing between devices takes place?

> The following of the scenario is then rather simple. Once the TLS channel has been opened, the User authenticates towards the AS through that TLS channel

This only ensures the one TLS channel between the client and AS is authenticated. If the client makes a request to an RS, it would be a different TLS channel because the RS is not necessarily at the same host as the AS. That new TLS session with the RS then does not carry the same level of trust as the one with the AS because the user did not authenticate within the TLS session with the RS.

Also this doesn't explain how the user can pick up a new device and move to that. You also keep coming back to this TLS session idea while also saying you don't want to bind authentication sessions to TLS.

**Comment by agropper** [comment]
I'm having a hard time following.

I can understand the case where the same RO uses four different client
instances to access the AS. But in an attenuated capabilities access token
model, the RO can transfer the capability to a RqP-1using some other
client. That RqP-1 can then transfer an attenuated capability to another
RqP-2 that has absolutely no relationship to the RO or the AS or the RS and
will bring whatever client they want to access the RS.

Am I talking about a completely different situation than the RO has 4
client instances?

**Comment by jricher** [comment]
The use of TLS in this way also doesn't work for something as simple as a client on a mobile device. In this: A client instance connects to an AS over a TLS channel which then gains trust. Then that same client instance jumps to another network and has to create a new TLS channel. Does the user need to re-authenticate because they moved their phone? With the trust model described above they would have to. Alternatively, you need a credential that stays with the client across TLS sessions, and that's what the client instance's key is.

**Comment by jricher** [comment]
@agropper yes, what you're describing is a completely different thing. Attenuated capabilities would be based on some other artifact (or set of artifacts and policies) from the RO to each of the RqP's, possibly including the ability to redelegate again. But none of this is carried in the access token, since the access token is for accessing the RS and not for identifying the RO or RqP itself. I'd recommend you start a new issue thread for that or talk about it more in #169

**Comment by Denisthemalice** [comment]
@aaronpk may be a phone conference would help to speed up the advancement of the discussion.

You wrote:

> How do you propose this sharing between devices takes place?

Since the solution I propose does not associate any key with any device, the User can choose any device (as long as he believes
that it does not include some spyware, virus and anything equivalent).

You wrote:

>  If the client makes a request to an RS, it would be a different TLS channel because the RS is not necessarily at the same host as the AS. That new TLS session with the RS then does not carry the same level of trust as the one with the AS because the user did not authenticate within the TLS session with the RS.

This is the second TLS channel opened between the Client with the RS. The level of trust is obtained through the use of the first access token sent over that TLS channel which must include (among other fields) a **_"bui"_** field (**_binding user identifier_**). Please refer to the full story of that field with the Cases 1, 2.1 and 2.2. What is important is how that "**_bui_**" field it must be checked by the RS depending upon its type.

**Comment by agropper** [comment]
Thanks for the confirmation. #169 seems fine as is and remains open.

**Comment by aaronpk** [comment]
> Since the solution I propose does not associate any key with any device, the User can choose any device

Yes, this is the case with using client instance keys as well.

> This is the second TLS channel opened between the Client with the RS. The level of trust is obtained through the use of the first access token sent over that TLS channel

This does not provide the assurance you are describing. The lack of client instance key in this case means any piece of software that is able to extract an access token would be able to use it at the RS masquerading as the legitimate user and client. Client instance keys prevent this from being possible. We have described this _in this thread_ before.

**Comment by Denisthemalice** [comment]
@jricher . You wrote:

> Then that same client instance jumps to another network and has to create a new TLS channel. Does the user need to re-authenticate because they moved their phone? With the trust model described above they would have to. Alternatively, you need a credential that stays with the client across TLS sessions, and that's what the client instance's key is.

The approach I propose works on fixed channels, which is what I experience  when I am at a fixed location. You are addressing a specific case. 

I am not an expert in mobile channels, but the alternative solution you are suggesting would need to be explored in more depth since it seems simpler to manage than managing fixed keys assigned to Users and Client instances.

**Comment by aaronpk** [comment]
> The approach I propose works on fixed channels, which is what I experience when I am at a fixed location. You are addressing a specific case.

Quite the opposite, fixed channels with unchanging IPs and consistent TLS channels are the exception. Mobile devices surpassed desktop years ago. Even within residential internet service it is not uncommon that the ISP may change the user's IP address, requiring a new TLS session.

**Comment by Denisthemalice** [comment]
@aaronpk We are cross-posting with a delay of a few minutes which is not very easy.
 
You wrote:

>    Since the solution I propose does not associate any key with any device, the User can choose any device

> Yes, this is the case with using client instance keys as well.

Maybe you should try to define the trust conditions you have in the current model between the AS, the User and the client instances (i.e. devices) used by the User. In the model I am proposing, the User can choose any device because **_the device does not contain any client instance key_**.

Then you wrote:

> The lack of client instance key in this case means any piece of software that is able to extract an access token would be able to use it at the RS masquerading as the legitimate user and client.

Remember the story between user1 and user2: user2 cannot not get advantage of an access token that has ben issued for user1. 

In the current document, if a user simply lends his smartphone to another person, it can immediately impersonate that person.
With the approach I propose, if a user simply lends his smartphone to another person, that other person cannot impersonate the owner of that smartphone (because he cannot authenticate to the AS).

**Comment by Denisthemalice** [comment]
@aaronpk . You wrote:
> Even within residential internet service it is not uncommon that the ISP may change the user's IP address, requiring a new TLS session.

For your information, I have the  same IP address for 3 years with the same ISP. You can pay to have a guaranteed fixed IP address 
(I didn't pay for it) , but in practice ISPs do not change IP addresses. If someone is going to choose a different ISP, it seems natural for me, that he should  terminate what he is doing before switching to another ISP.

**Comment by fimbault** [comment]
@Denisthemalice it's not exactly what I said. I explained we do need a key per client instance. And this may have a consequence on the user flow, just like it has today for password protected websites. Note that this doesn't invalidate the architecture put forward in gnap, it just makes it mandatory to manage the keys somehow.

**Comment by Denisthemalice** [comment]
@fimbault . You wrote :

> it's not exactly what I **_said_**. I explained we do need a key per client instance.


As long as the trust conditions will not be **written**,  we will continue to argue for ever.

**Comment by Denisthemalice** [comment]
A naive question: 

If I lend my smartphone to some one else for about 10 minutes, how difficult it is to extract and export a client instance key stored
in a smartphone supporting either Android or IOS ?

**Comment by fimbault** [comment]
You cannot simply extract, export, view private key material from KMS. Then of course it all depends on the actual execution environment that underlies those stores.

**Comment by jricher** [comment]
If you're using the device enclave to manage your keys, then it would be very difficult to nearly impossible, especially within that timeframe. If it's a software-based key, then it would still be reasonably difficult within that timeframe. 

Please understand that the requirement for each client instance to have a key does NOT necessitate a pre-registration of that key with the AS (or RS) ahead of time. Mobile devices can generate keypairs on-device and keep them in storage enclaves for use -- think of it like a consumer-grade HSM. GNAP has been designed to allow for this type of key introduction from the start, with the client's "key" section in the request.

**Comment by Denisthemalice** [comment]
@jricher . You wrote:

> Please understand that the requirement for each client instance to have a key does NOT necessitate a pre-registration of that key with the AS (or RS) ahead of time.

Looking at the API parameters to understand a key management architecture is all what we have today, How would you handle 
key registration "on the fly" ? 

I have not taken a close look at the client RS architecture in terms of client instance keys, but reading your sentence above  which includes "(and RS)" do you mean that clients must also register their keys to each RS ?

**Comment by Denisthemalice** [comment]
@fimbault . You wrote:

> You cannot simply extract, export, view private key material from KMS. Then of course it all depends on the actual execution environment that underlies those stores.

With the current architecture, the AS cannot know how good (or how bad) the client is protecting the client private keys. 
So I would guess, that the AS will accept the client public key simply based on a declaration from the user.
This is rather insecure.

**Comment by jricher** [comment]
@Denisthemalice how can you claim that accepting a key based on a user declaration is insecure, but accepting a TLS socket connection on the same basis is somehow more secure? And this is not even based IP address, since TLS connections get set up and torn down all the time from one IP. Nearly every HTTP call is likely over a new TLS connection between the same parties.

**Comment by Denisthemalice** [comment]
@jricher . You wrote:

> how can you claim that accepting a key based on a user declaration is insecure, 

Since the registration process is left currently undescribed, this is an assumption for the moment. 
Now, if you can explain why it is secure, your explanations will be welcomed.

You also wrote:

> but accepting a TLS socket connection on the same basis is somehow more secure? 

I have not made any comparison, since no comparison can be made: when a TLS session is being opened, the server is kept ignorant of the characteristics of the client (and of the user).

**Comment by Denisthemalice** [comment]
@aaronpk . Earlier this done, you wrote:

> [Denis]   What is not explained in the current specification, is that each user must contact the AS in order to be first associated with one client instance key.
> 
> From the introduction: "This protocol allows a piece of software, the client instance, to request delegated authorization to resource servers", and the term "client instance" is described in the section defining the roles.
> 
> This seems to describe the situation accurately to me. Do you have a suggestion for making this clearer?

Since the words "user" or "end-user" are not present in this text, how do you want a reader to understand that there is a link between users and client instance keys ? 

My suggestion is the following: a section explaining the **_trust conditions_** between the various actors has been requested about three months ago and is still not present. In additionn to this static description, it would be useful to explain how the system can be "bootstrapable". I believe that the top priority should to write such a section.

**Comment by aaronpk** [comment]
> Since the words "user" or "end-user" are not present in this text, how do you want a reader to understand that there is a link between users and client instance keys ?

It is in the rest of the introduction paragraph I quoted, so please read that first.

> My suggestion is the following: a section explaining the trust conditions between the various actors has been requested about three months ago and is still not present.

Anyone in this working group is welcome to propose text for the spec, so you have had 3 months to provide suggestions for this section as well. Complaining that others have not done work that you want is not constructive.

**Comment by fimbault** [comment]
Trust based on public key cryptography is pretty established. Including how to protect against vulnerabilities and key theft.
There have also been improvements lately, due to their heavy use in blockchain protocols. In our space of identity management, keri brings additional security as it not only provides a keypair, but also a full trail of events (and therefore becomes much harder to forge). 

After careful review of your proposal, I believe the core GNAP assumptions remain well suited to the task, as long as we make sure keys are well managed (which is planned).

**Comment by Denisthemalice** [comment]
@aaronpk : I read both sections. I have been unable to find some sentences that allow to understand that a user must register his client instance key(s) with the AS and that an AS will then associate that User with this/these client instance key(s). This level of understanding has only been possible ... yesterday, because Fabien provided these explanation to me on the phone.

It is up to the people that design a protocol to write the trust conditions. I could only understand them ... yesterday, and not three months ago. In addition, you are best placed to write them, otherwise you might complain that they are inaccurate, incomplete or even wrong. 

@fimbault : Trust is not only **_based on public key cryptography_**. 

May I recall, once again, that the template to express a trust condition is the following:

> an entity A is trusting an entity B for something or some kind of behavior.

You wrote:

> I believe the core GNAP assumptions remain well suited to the task, as long as we make sure keys are well managed 
(which is planned).

The use of client instance keys can be avoided with my proposal which has the major advantage of avoiding **_client instance key registration_** (which is not mentioned in the current draft) and also **_client instance key rotation_** and also **_client instance key revocation_** (which is not mentioned in the current draft). It also solves the so-called **_mix-up attack_** issue.

In terms of **_management_**, this means less management people for the AS which is important in terms of **_deployment costs_**. 

In term of **_ease of use_**, this means less burden for the users since they will need to register only once with the AS whatever the number of devices they will use. In addition, the AS will be kept ignorant of the number of devices used by each user and of their characteristics, which is a nice in terms of **_privacy_** for the users.

In terms of **_mobility_**, this means that users will be able to use any kind of device as long as they can trust it to be free of viruses, Trojan horse and the like. When they will use a new device, they will be able to use it immediately without the burden to register their new device.

**Comment by fimbault** [comment]
@Denisthemalice in gnap, the association is not with the user, but with the client instance. What you propose is different (but don't impose it on gnap when you ask questions!). If I understand correctly, you want a user attribute bound to a session, the attribute being able to identify who is calling. While that's interesting, I'm concerned by whether this can even be implemented, it doesn't seem so obvious to me.

**Comment by aaronpk** [comment]
> The use of client instance keys can be avoided with my proposal which has the major advantage of avoiding client instance key registration (which is not mentioned in the current draft)

Client instance key registration is described in [section 2.3, "Identifying the Client Instance"](https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-06.html#name-identifying-the-client-inst)

> It is up to the people that design a protocol to write the trust conditions. I could only understand them ... yesterday, and not three months ago. 

Denis, you have had more than three months to read the drafts and work with us on understanding what is currently written. Instead, you came in to the group demanding that the editors support your own model. You have continued to repeat the same arguments over and over despite not making any actual contributions to the draft and not making any attempt to understand the current model, and failing to get any support from others in the group. It is not constructive to participate in the group by making demands this way.

At this point I believe you would be better off writing your own individual draft and bringing it to [secdispatch](https://datatracker.ietf.org/wg/secdispatch/about/).

**Comment by Denisthemalice** [comment]
@aaronpk . You wrote:

> Client instance key registration is described in section 2.3, "Identifying the Client Instance";

Would you be able to tell me **_when reading this text_** what is the relationship between a user (or an end-user) and a client instance ?

The words "user" or "end-user" are not present in this section !

On March 12 th, **more than 4 months ago**, I raised the issue #214 which was about trust relationships.
I proposed some text about the trust relationship.

It is very unfortunate that **you have not participated to this thread** because I had to wait until **yesterday** to understand just a little bit of the trust relationships concepts(thanks to the help of Fabien). Since these trust relationships concepts are still not written at this time for a model supporting client instance keys, I am not even sure that I understand the whole of it. Please read issue #214 and please write the trust relationships that apply to your model.

I have just written the trust relationships that apply to my model. This topic is now covered under the issue #301.
I suggest that you read this issue before writing the trust relationship that apply to your model.

**Comment by aaronpk** [comment]
The relationship between the client and end user is defined in the [introduction section](https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-07.html#name-roles)

> Client: application operated by an end-user that consumes resources from one or several RSs

**Comment by Denisthemalice** [comment]
@aaronpk : I have a few comments on your statement:

First of all, the quoted sentence allows to contradict what @jricher said yesterday during the interim WG meeting.

@jricher said that the client could have no end-user for operating it, which is a case that is not possible with the quoted sentence.

Secondly, this is not a trust relationship.

At this point of time, the following trust relationships are still left undefined in draft -07:

-  between the end-user and a client instance, and
-  between an end-user and an AS.

**Comment by fimbault** [comment]
@Denisthemalice the relationship between an end-user and a client instance is already defined. As already explained, I initially decided not to describe the trust relationship between end-user and AS because it requires a bit more thought (interactions are varied, and it's going through technical agent such as a browser). But I'll do that.
